internal/infrastructure/mongo/config.go

```diff
@@ -16,30 +22,103 @@ var upsert = true
 
 type configRepo struct {
 	client *mongodoc.ClientCollection
+	lockID *uuid.UUID
+	lock   sync.Mutex
 }
 
+type ConfigDoc struct {
+	Migration int64
+	Auth      *config.Auth
+	lock      *uuid.UUID
+}
+
+var (
+	ErrLoadingLockedConfig = errors.New("loading locked config")
+	ErrConfigNotLocked     = errors.New("trying to save not locked config")
+)
+
 func NewConfig(client *mongodoc.Client) repo.Config {
 	return &configRepo{client: client.WithCollection("config")}
 }
 
 func (r *configRepo) Load(ctx context.Context) (*config.Config, error) {
-	cfg := &config.Config{}
-	if err := r.client.Collection().FindOne(ctx, nil).Decode(cfg); err != nil {
-		if errors.Is(err, mongo.ErrNoDocuments) {
+	r.lock.Lock()
+	defer r.lock.Unlock()
+	numOfTries := 2
+
+	for i := 1; i <= numOfTries; i++ {
+		cfg, err := r.loadFromDB(ctx)
+		if err == nil {
 			return cfg, nil
 		}
+		if !errors.Is(err, ErrLoadingLockedConfig) {
+			return nil, err
+		}
+		time.Sleep(time.Duration(rand.Intn(1000)+1000) * time.Millisecond)
+	}
+
+	log.Errorf("failed to load config after %d tries.\n", numOfTries)
+	return nil, ErrLoadingLockedConfig
+}
+
+func (r *configRepo) loadFromDB(ctx context.Context) (*config.Config, error) {
+	cfg := &ConfigDoc{}
+	if r.lockID == nil {
+		lockID := uuid.New()
+		r.lockID = &lockID
+	}
+
+	if err := r.client.Collection().FindOneAndUpdate(ctx,
+		bson.M{"$or": bson.A{
+			bson.M{"lock": bson.M{"$exists": false}},
+			bson.M{"lock": bson.M{"$eq": r.lockID}},
+		}},
+		bson.M{"$set": bson.M{"lock": r.lockID}},
+	).Decode(cfg); err != nil {
```

Comment:
Using upsert leads into config duplication in db, I tried this suggestion and it cause a duplication when tiring to load a locked config.
anyway I updated the implementation but this assumes the the config should be in the db other wise it will not created automatically